text/0011-gem-checksum-verification.md

```diff
@@ -0,0 +1,217 @@
+Commands that rely on checksums for verification will output a message or warning when checksums are missing or mismatched.
+
+Example:
+
+```
+$ bundle install
+Bundle complete! 88 Gemfile dependencies, 256 gems now installed.
+Use `bundle info [gemname]` to see where a bundled gem is installed.
+Use `bundle lock` to add missing checksums to Gemfile.lock.
+```
+
+Running `bundle update` or `bundle add` will record the checksum from the source (e.g. rubygems.org) into the Gemfile.lock, if it is available.
+If a checksum is not available from the source because the source does not provide such info (e.g. private gemservers) then a checksum will be created during install using the .gem file.
+If a checksum can't be created because the source is a path or git source, then only the gems name and version will be recorded with a blank checksum.
+
+If you want to ensure that the bundled environment contains only gems matching the checksums in the lockfile, run `bundle pristine`.
+The `bundle pristine` command installs every gem fresh from a newly downloaded .gem source.
+The pristine install will trigger computation and comparison of the generated SHA256 checksum with the checksum stored in the lockfile.
+
+Example:
+
+```
+$ bundle pristine
+Installing rake 13.0.6
+…
+Installing rspec 3.12.0
+42 gems without checksums.
+Use `bundle lock` to add checksums to Gemfile.lock.
+$ bundle lock
+Writing lockfile to path/to/Gemfile.lock
+$ bundle pristine
+Installing rake 13.0.6
+…
+Installing rspec 3.12.0
+$
+```
+
+When bundler installs a gem from a `.gem` file, it computes the SHA256 checksum of the file.
+If an existing checksum is available from the lockfile or the remote source, it will be compared with the computed checksum at install time.
+If the checksums do not match, an error is generated and installation is halted.
+If no checksum is recorded in the lockfile, the computed checksum is saved to the lockfile where future installs can verify that the same gem is installed.
+
+Example:
+
+```
+$ bundle install
+Installing rake 13.0.6
+Bundler found mismatched checksums. This is a potential security risk.
+  rake (13.0.6) sha256-2222222222222222222222222222222222222222222222222222222222222222
+    form the lockfile CHECKSUMS at Gemfile.lock:21:17
+  rake (13.0.6) sha256-814828c34f1315d7e7b7e8295184577cc4e969bad6156ac069d02d63f58d82e8
+    from the gem at path/to/rake-13.0.6.gem
+
+  To resolve this issue you can either:
+    1. remove the gem at path/to/rake-13.0.6.gem
+    2. run `bundle install`
+  or if you are sure that the new checksum from the lockfile CHECKSUMS at Gemfile.lock:21:17 is correct:
+    1. remove the matching checksum in Gemfile.lock:21:17
+    2. run `bundle install`
+
+To ignore checksum security warnings, disable checksum validation with
+  `bundle config set --local disable_checksum_validation true`
```

Comment:
IMO no, if we want a way to avoid putting checksums in the lock file (which im also not sure we want), it should be under a different flag

Reply:
I'm also not sure we want it, but I reckon it could be useful for a graceful migration, given that the setting already exists, even if we'll eventually remove the flag.

It'd be interesting to check why we added this flag in the first place.

Reply:
Good question. It was there when I accidentally stole this work from @segiddins. It's a failsafe in case we really mess someone up or they like using hacked and corrupted gems and wish to continue doing so.